### PR TITLE
Fix error that bugs Keras during multi-gpu with batch normalization

### PR DESCRIPTION
**Pull request by owsd**
Here are the issues that lead to this bug:
https://github.com/keras-team/keras/issues/10008
https://github.com/keras-team/keras/issues/10490

If you use `multi_gpu_model` in Keras, it causes an error.  I tracked this to the following stack:

```
  File "/home/sean/env3/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py", line 1858, in normalize_batch_in_training
    if not _has_nchw_support() and list(reduction_axes) == [0, 2, 3]:
  File "/home/sean/env3/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py", line 291, in _has_nchw_support
    explicitly_on_cpu = _is_current_explicit_device('CPU')
  File "/home/sean/env3/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py", line 266, in _is_current_explicit_device
    device = _get_current_tf_device()
  File "/home/sean/env3/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py", line 247, in _get_current_tf_device
    g._apply_device_functions(op)
  File "/home/sean/env3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 4259, in _apply_device_functions
    op._set_device(device_spec.function(op))
  File "/home/sean/env3/lib/python3.6/site-packages/tensorflow/python/framework/device.py", line 306, in _device_function
    current_device = DeviceSpec.from_string(node_def.device or "")
  File "/home/sean/env3/lib/python3.6/site-packages/tensorflow/python/framework/device.py", line 231, in from_string
    return DeviceSpec().parse_from_string(spec)
  File "/home/sean/env3/lib/python3.6/site-packages/tensorflow/python/framework/device.py", line 149, in parse_from_string
    splits = [x.split(":") for x in spec.split("/")]
AttributeError: 'DeviceSpec' object has no attribute 'split'
```

The fix is that for some reason, the node_def.device from `device.py` is being passed in as a `DeviceSpec` instance, not a string as expected.  However, I noticed a couple of lines above that, there is a check to see if `spec` is a `DeviceSpec` instance.  I went ahead and added this check to this line as well, and it fixes the issue.

### CODE REVIEW

**Comment by owsd** [comment]
I signed it!

**Review by protoget — CHANGES_REQUESTED** [review]

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ymodak** [comment]
@owsd Can you please resolve the merge conflicts and make requested changes to move ahead with this PR? Thanks!

**Comment by rthadur** [comment]
@owsd  gentle ping to resolve the merge conflicts

**Comment by dkuegler** [comment]
I am facing the same problem on TF 1.13.1 and keras 2.2.4.

I have tracked it one step further to:
`self._device_function_stack.peek_objs()` in `_apply_device_functions(self, op)` returning a `<tensorflow.python.framework.device.DeviceSpec object at 0x7fc3ab9259e8>` as second in iteration.

`python3.6/site-packages/keras/backend/tensorflow_backend.py in _get_current_tf_device()
    246     op = _TfDeviceCaptureOp()
    247     print('K.', op.device)
--> 248     g._apply_device_functions(op)
    249     return op.device
    250 

python3.6/site-packages/tensorflow/python/framework/ops.py in _apply_device_functions(self, op)
   4244       if device_spec.function is None:
   4245         break
-> 4246       op._set_device(device_spec.function(op))
   4247     op._device_code_locations = self._snapshot_device_function_stack_metadata()
   4248     # pylint: enable=protected-access

python3.6/site-packages/tensorflow/python/framework/device.py in _device_function(node_def)
    313     def _device_function(node_def):
    314       print("-", node_def.device, "-")
--> 315       current_device = DeviceSpec.from_string(node_def.device or "")`

From my reading this does not seem to be dependent on the way keras calls tensorflow, but dependent on the population of the device_function_stack of the graph.

Unfortunately, I can't really determine how to reproduce the error, but I tried to multi_gpu_model() a pretrained DenseNet121 from keras.

**Comment by dkuegler** [comment]
I cant really provide a better minimal working example right now, but maybe this helps.

[tensorflow_23197_minimal.py.txt](https://github.com/tensorflow/tensorflow/files/3026051/tensorflow_23197_minimal.py.txt)

**Comment by dkuegler** [comment]
However, this Pull Requests solution seems to fix things.

File: tensorflow/python/framework/device.py [hunk 1]
```diff
@@ -283,7 +283,8 @@ def merge_device(spec):
   if not isinstance(spec, DeviceSpec):
     spec = DeviceSpec.from_string(spec or "")
   def _device_function(node_def):
-    current_device = DeviceSpec.from_string(node_def.device or "")
+    current_device = node_def.device if isinstance(node_def.device, DeviceSpec) \
```
[the remark this answers is not in the document]
Comment (WangShuaixian) [reply tensorflow/python/framework/device.py]: I sloved it by 
open  /tensorflow/python/framework/device.py
change line314  `current_device = DeviceSpec.from_string(node_def.device or "")` to 
`current_device = node_def.device if isinstance(node_def.device, DeviceSpec) else DeviceSpec.from_string(node_def.device or "")`

**Pull request closed without merging**